### PR TITLE
Improve release process with slack notifications and automated branching

### PR DESCRIPTION
**Pull request by kalanchan**
This will PR will do the following:
1. push a notification to a slack channel informing us that the release is successful
2. calculates the next prerelease version, and provides a preview of the changes going into the next version (stopgap solution for GTM and PMM)
3. creates the next backport label
4. creates the next prerelease branch and pushes it to main

It will only create new prerelease branches and backport labels when there is a new minor version, patch increments will be skipped. 

The slack notification will look something like:

CleanShot 2024-11-28 at 15 48 33

This will reduce the cognitive load on the release captain to always remembering to cut the next release branches, reducing the chance of error. The release captain just needs to make sure that they merge changelog and version bumps first. 

next up is doing this for the pre-release and JetBrains repos

## Test plan
N/A - tested in private [slack channel](https://sourcegraph.slack.com/archives/C04JD75GY2U)

### CODE REVIEW

File: .github/workflows/vscode-stable-release.yml [hunk 1]
```diff
@@ -68,3 +68,76 @@ jobs:
...
+            })      
+      - name: Create and push next release branch
+        if: ${{ !env.SKIP_CREATION }}
+        run: |
+          git checkout -b ${{ env.NEXT_RELEASE_BRANCH }}
+
+          if ! git push origin ${{ env.NEXT_RELEASE_BRANCH }}; then
+            echo "Failed to push the next release branch ${{ env.NEXT_RELEASE_BRANCH }} to origin"
+            exit 1
+          fi
```
Comment (kalanchan) [inline .github/workflows/vscode-stable-release.yml:143]: @valerybugakov what do you think about automating the next backport label and release branch? I think it make the release process less annoying but will require everyone to be a bit more diligent on following the playbook steps

Comment (valerybugakov) [reply .github/workflows/vscode-stable-release.yml:143]: > everyone to be a bit more diligent on following the playbook steps

This is the result we’re aiming for. Let’s give it a try and see if it turns out to be a net positive or just adds more confusion.

File: .github/workflows/vscode-stable-release.yml [hunk 2]
```diff
@@ -68,3 +68,76 @@ jobs:
...
+              echo "❌ Unable to post message to slack, got:"
+              echo "--- raw body ---"
+              echo "$body"
+              echo "--- raw body ---"
+              exit 1
+            else
+              echo "Posted to slack."
+            fi
+          done
+      - name: Check if branch or label exists
```
Comment (valerybugakov) [inline .github/workflows/vscode-stable-release.yml:106]:
```suggestion
      - name: Check if the next release branch and backport label exist
```

File: .github/workflows/vscode-stable-release.yml [hunk 3]
```diff
@@ -68,3 +68,76 @@ jobs:
           asset_path: ./vscode/dist/cody.vsix
           asset_name: cody-vscode-${{ env.EXT_VERSION }}.vsix
           asset_content_type: application/zip
+      - name: Determine version numbers
+        run: |
+          tag="${{ env.EXT_VERSION }}"
+          version_anchor=$(echo $tag | sed 's/[^0-9]//g')
+          major=$(echo $tag | sed 's/\([0-9]*\)\.\([0-9]*\)\.\([0-9]*\)/\1/')
+          minor=$(echo $tag | sed 's/\([0-9]*\)\.\([0-9]*\)\.\([0-9]*\)/\2/')
```
Comment (valerybugakov) [inline .github/workflows/vscode-stable-release.yml:76]: I would prefer to have this logic (and all of the following steps) in Typescript for easier debugging and the opportunity to cover unit tests. We have some scripts in `vscode/src/release.ts`. It's not a strong stance, but it should make our lives easier when something goes wrong. I'm curious what @dominiccooney thinks about it.

File: .github/workflows/vscode-stable-release.yml [hunk 4]
```diff
@@ -68,3 +68,76 @@ jobs:
           asset_path: ./vscode/dist/cody.vsix
           asset_name: cody-vscode-${{ env.EXT_VERSION }}.vsix
           asset_content_type: application/zip
+      - name: Determine version numbers
+        run: |
+          tag="${{ env.EXT_VERSION }}"
+          version_anchor=$(echo $tag | sed 's/[^0-9]//g')
```
[the remark this answers is not in the document]
Comment (kalanchan) [reply .github/workflows/vscode-stable-release.yml:74]: yeah i think thats fine, just usng `version_anchor` for the changelog url

for example `https://github.com/sourcegraph/cody/blob/main/vscode/CHANGELOG.md#1460` 1460 would be the version anchor

File: .github/workflows/vscode-stable-release.yml (hunk 3 above)
Comment (kalanchan) [reply .github/workflows/vscode-stable-release.yml:76]: is there an easy to way to run git commands and pull in github secrets? one of the advantages of using GHA is the github integration and built in support for steps

Comment (valerybugakov) [reply .github/workflows/vscode-stable-release.yml:76]: > is there an easy to way to run git commands

Yes, we can [spawn](https://github.com/sourcegraph/cody/blob/9d4b1b7a1d0b9b0a5fde008a90be1e639870d122/vscode/test/e2e/utils/gitWorkspace.ts#L55-L58) a child process to run commands for us.

> and pull in github secrets

We can set all the required secrets as env variables which would be accessible in the script.

> one of the advantages of using GHA is the github integration and built in support for steps

Agree, it looks pretty straightforward for now. But if these actions get more complex over time, debugging and fixing issues will become exponentially harder.

**Review by valerybugakov — APPROVED** [review]
Thank you! 🤩

File: .github/workflows/vscode-stable-release.yml (hunk 3 above)
Comment (kalanchan) [reply .github/workflows/vscode-stable-release.yml:76]: let's try this for now, if the automated branching becomes too messy we can remove that and just use the notifications part

**Pull request merged**